### ISSUE TITLE
Improve data resiliency guarantees for writes to ElasticSearch

### ISSUE THREAD

**Issue by deepthidevaki** [camunda/camunda#21535]
**Description**

ElasticSearch do no have the same guarantees for data resiliency as in Zeebe. That means, even when we have configured 1 or more replicas for the shards in ElasticSearch
1. There is a possibility of unavailability if primary is down
2. There is a possibility of partial dataloss if primary loses its data

Here is the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-replication.html) for further reference.

Summary:
1. There are primaries and in-sync replicas.
2. Client requests are acknowledged after it is written in both primary and in-sync replicas.
3. However, during failure scenarios if an in-sync replica is not reachable or responds, primary will mark the in-sync replica as stale and acknowledge the write to the client without the confirmation from the in-sync replica.
4. As a result, it is possible that there is only on copy of the data at some point. 
5. Elasticsearch gurantees no data loss by not allowing the stale replicas to be promoted as primary.
6. In that case, if the primary is down, that shard is unavailable for writes. 
7. Also in that case, if the data in primary is lost (due to physical hardware issues or natural disasters), then the index suffers from partial data loss.

> This means that the master knows that the primary is the only single good copy. We are therefore guaranteed that the master will not promote any other (out-of-date) shard copy to be a new primary and that any operation indexed into the primary will not be lost. Of course, since at that point we are running with only single copy of the data, physical hardware issues can cause data loss. 

To avoid this case we can in addition use [index.writes.wait_for_active_shards](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-index_.html#index-wait-for-active-shards) configuration. However, this also does not 100% guarantee that atleast 2 nodes (primary and 1 replica) has the latest data.
> It is important to note that this setting greatly reduces the chances of the write operation not writing to the requisite number of shard copies, but it does not completely eliminate the possibility, because this check occurs before the write operation commences. Once the write operation is underway, it is still possible for replication to fail on any number of shard copies but still succeed on the primary. The _shards section of the write operation’s response reveals the number of shard copies on which replication succeeded/failed.

As a mitigation, we could check the response before considering the write as successful. 

*Goal*

* Make `index.write.wait-for-active-shards` configurable from the apps. 
* Decide a sensible defaults if shards are replication. If they are not replicated, this value should be 1 and it is not useful.
* Handle the case where the writes are still acknowledged even if it was not applied to the minimum number of replicas. 

Note that, this applies to all apps that writes to ElasticSearch. 
Do not know how this impacts the interaction with OpenSearch.

**Comment by npepinpe** [comment]
ZDP Triage:
- You mentioned it may be possible to diagnose this issue even without a mitigation. Is that correct?

**Comment by deepthidevaki** [comment]
> * You mentioned it may be possible to diagnose this issue even without a mitigation. Is that correct? If that's the case, I would put this in the back log.

Yes. Kind of. ES doesn't automatically promote a stale replica to primary. I believe what happens is that there is no primary and we can somehow observe this in ElasticSearch health status or metrics. I'm not sure how exactly though. If this is affecting zeebe-indices, we would see errors in zeebe exporter. 

From the docs:

> What happens if there are no replicas?
> 
> This is a valid scenario that can happen due to index configuration or simply because all the replicas have failed. In that case the primary is processing operations without any external validation, which may seem problematic. On the other hand, the primary cannot fail other shards on its own but request the master to do so on its behalf. This means that the master knows that the primary is the only single good copy. **We are therefore guaranteed that the master will not promote any other (out-of-date) shard copy to be a new primary and that any operation indexed into the primary will not be lost**. Of course, since at that point we are running with only single copy of the data, physical hardware issues can cause data loss.

**Comment by npepinpe** [comment]
I will put it in the backlog _for us_ as this is a cross functional issue. We should put it front and center when it comes to the advanced and critical offering however. I'll bring it up there.

**Comment by RomanJRW** [comment]
A few notes:
- OpenSearch has the same equivalent setting
- We recently decided to increase the default replica count from `0` to `1`, so this becomes a little more relevant in a way. Some thoughts:
  - We could set this to 1 by default, but allow it to be configured. For multi-node clusters with >0 replicas, we could set this to 2  (or `all`?) in the Helm Chart and operator
  - We should only do this if we think the risk is significant and worth the inevitable performance hit on write (which we need to evaluate up front)

Moving this to the backlog with a prio and complexity estimate
